kv/memdb/arena.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memdb
+
+import "math"
+
+type arenaAddr struct {
+	blockIdx    uint32
+	blockOffset uint32
+}
+
+func (addr arenaAddr) isNull() bool {
+	return addr.blockIdx == 0 && addr.blockOffset == 0
+}
+
+func newArenaAddr(idx int, offset uint32) arenaAddr {
+	return arenaAddr{
+		blockIdx:    uint32(idx) + 1,
+		blockOffset: offset,
+	}
+}
+
+const (
+	nullBlockOffset = math.MaxUint32
+	maxBlockSize    = 128 << 20
+)
+
+type arena struct {
+	blockSize int
+	availIdx  int
+	blocks    []*arenaBlock
+}
+
+func newArenaLocator(initBlockSize int) *arena {
+	return &arena{
+		blockSize: initBlockSize,
+		blocks:    []*arenaBlock{newArenaBlock(initBlockSize)},
+	}
+}
+
+func (a *arena) getFrom(addr arenaAddr) []byte {
+	return a.blocks[addr.blockIdx-1].getFrom(addr.blockOffset)
+}
+
+func (a *arena) alloc(size int) (arenaAddr, []byte) {
+	if size > a.blockSize {
+		// Use a separate block to store entry which size larger than specified block size.
+		blk := newArenaBlock(size)
+		addr := newArenaAddr(len(a.blocks), 0)
+		a.blocks = append(a.blocks, blk)
+		return addr, blk.buf
+	}
+
+	for {
```

Comment:
if size > maxBlockSize, this is a dead loop and lead to OOM eventually?

Reply:
`a.blockSize <= maxBlockSize` and `alloc` checked `if size > a.blockSize` at L57. So `size > maxBlockSize` never step into this loop.